### ISSUE TITLE
GazeR functions exceed my working space limit: how to downsample

### ISSUE THREAD

**Issue by juthzi** [dmirman/gazer#7]
Dear GazeR-Team,

I have trouble processing my pupil data with some of the GazeR functions. I run into working space limit problems.

Therefore, I wanted to downsample the files (they have been created with 1000 Hz) to 250 Hz for a start. In the GazeR manual you describe a way to downsample using "downsample_gaze", but for me, this step occurs too late in the script, so that I run into problems beforehand. Do you know, if there is a way to downsample the edf files when they are initially parsed(edf) or imported (csv) into R? Or can I simply use the "downsample_gaze" at an earlier point in the script without messing other things up?

For explanation:
The problem is not so much the reading in of the files, but running the functions following after that, so far: behave_pupil() and baseline_correction_pupil_msg().

I guess I could run these functions on individual files or on only a couple of files. However, I am reluctant to do this, because this means, that I have to run the whole pipeline for an individual file and afterwards concatenate the files, which always creates new sources of errors (I am not that firm with R yet, or the package).

Thank you for your advice,
Juliane

**Comment by jgeller112** [comment]
You can use the below function to downsample your data before you begin the gazeR routines:

```
#samplerate=current
#newrate=new sampling rate
downsamle_pupil_prior <- function(data, SamplingRate = NULL, NewRate = NULL) {
  
  # Overwrite Pupil_Previous
  data <- data %>% arrange(subject, trial, time) %>%
    mutate(pupil_previous = pupil , gaze_x_previous = x, gaze_y_previous = y)
  
  if(is.null(SamplingRate)){
    stop("Please supply the sampling rate!")
  }
  if(is.null(NewRate)){
    stop("Please supply the new rate!")
  }
  
  sampleinterval <- 1000/data$time[2]-data$time[1]
  
  binsize <- 1000/NewRate
  
  
  data$DS <- data$time %/% binsize
  data$DS <- data$DS * binsize
  
  tmp <- data %>% group_by(subject, trial, DS) %>%
    mutate(pupil = stats::median(pupil_previous, na.rm = TRUE),
           x = stats::median(gaze_x_previous, na.rm = TRUE), y = stats::median(gaze_y_previous, na.rm = TRUE)) %>%
    filter(time %in% DS) %>%
    ungroup() %>% 
    select(-DS, -pupil_previous, -gaze_x_previous, -gaze_y_previous)
  
  return(droplevels(ungroup(tmp)))
}
```

**Comment by juthzi** [comment]
Thank you for your answer! I tried doing that for my data:

`pupil_files_test_downsampled <- downsamle_pupil_prior(data = pupil_files_test, SamplingRate = 1000, NewRate = 250)`

But even though I definitely used 1000 Hz for sampling, I get the following output:

`Error in downsamle_pupil_prior(data = pupil_files_test, SamplingRate = 1000,  : 
  Sampling rate provided does not match the sampling rate of the data.
 Please check the sampling rate using the function check_samplingrate()!
Called from: downsamle_pupil_prior(data = pupil_files_test, SamplingRate = 1000, 
    NewRate = 250)`

When trying to use the function check_samplingrate(), it does not exist in my libraries. In which package should the function be? 
Or do you think the reason for the error could be something else?

**Comment by jgeller112** [comment]
Can you please post the function you ran here—that error message is not in the above function

**Comment by juthzi** [comment]
Sure, I used the one proposed above:

```
downsamle_pupil_prior <- function(data, SamplingRate = NULL, NewRate = NULL) {
  
  data <- data %>% arrange(subject, trial, time) %>%
    mutate(pupil_previous = pupil , gaze_x_previous = x, gaze_y_previous = y)
  
  if(is.null(SamplingRate)){
    stop("Please supply the sampling rate!")
  }
  if(is.null(NewRate)){
    stop("Please supply the new rate!")
  }
  
  sampleinterval <- 1000/data$time[2]-data$time[1]*1000
  
  if(SamplingRate != 1000/sampleinterval){
    stop("Sampling rate provided does not match the sampling rate of the data.\n Please check the sampling rate using the function check_samplingrate()!")
  }
  
  binsize <- 1000/NewRate
  
  
  data$DS <- data$time %/% binsize
  data$DS <- data$DS * binsize
  
  tmp <- data %>% group_by(subject, trial, DS) %>%
    mutate(Pupil = stats::median(pupil_previous, na.rm = TRUE),
           Gaze_X = stats::median(gaze_x_previous, na.rm = TRUE), Gaze_Y = stats::median(Gaze_Y_Previous, na.rm = TRUE)) %>%
    filter(time %in% DS) %>% ungroup() %>% select(-DS, -pupil_previous, -gaze_x_previous, -gaze_y_previous)
  
  return(droplevels(ungroup(tmp)))
}

pupil_files_test_downsampled <- downsamle_pupil_prior(data = pupil_files_test, SamplingRate = 1000, NewRate = 250)

```

**Comment by jgeller112** [comment]
That’s not the function I have up there:

‘’’
downsamle_pupil_prior <- function(data, SamplingRate = NULL, NewRate = NULL) {

  data <- data %>% arrange(subject, trial, time) %>%
    mutate(pupil_previous = pupil , gaze_x_previous = x, gaze_y_previous = y)
  
  if(is.null(SamplingRate)){
    stop("Please supply the sampling rate!")
  }
  if(is.null(NewRate)){
    stop("Please supply the new rate!")
  }
  
  sampleinterval <- 1000/data$time[2]-data$time[1]
  
  binsize <- 1000/NewRate
  
  
  data$DS <- data$time %/% binsize
  data$DS <- data$DS * binsize
  
  tmp <- data %>% group_by(subject, trial, DS) %>%
    mutate(pupil = stats::median(pupil_previous, na.rm = TRUE),
           x = stats::median(gaze_x_previous, na.rm = TRUE), y = stats::median(gaze_y_previous, na.rm = TRUE)) %>%
    filter(time %in% DS) %>%
    ungroup() %>% 
    select(-DS, -pupil_previous, -gaze_x_previous, -gaze_y_previous)
  
  return(droplevels(ungroup(tmp)))
}

‘’’

**Comment by juthzi** [comment]
Great, thank you very much for editing the function, it works now!

Unfortunately, I run into the problem, that some rows including (important) messages are also deleted. I need to keep those for analysis. Do you have an idea how to best tackle this issue?

Otherwise, maybe you have an opinion on these ideas?
a) store the "message" column before downsampling and later add them again to the downsampled data by transfering the message to the next-nearest sample.
b) during downsampling: skip a row if it contains an entry in the "message" column. However then, there would be slight inaccuracies in the data.

**Comment by jgeller112** [comment]
Sadly, I do not know of a good solution for this. When merging the raw EDF files there are already some slight inaccuracies in merging the messages with the raw data--doing this again would add further variance that I am not comfortable with. If you want to take a stab at doing this and contribute something it would definitely be welcome. 

As an alternative, you can open the EDFs in Data Viewer and downsample that way. This will preserve your messages. 

J

**Comment by jgeller112** [comment]
@tjmahr may be able to offer a solution to this?

**Comment by tjmahr** [comment]
try reading the csv with vroom? it only reads in data from a file when it is needed.

https://cran.r-project.org/web/packages/vroom/vignettes/vroom.html

**Comment by jgeller112** [comment]
@tjmahr any suggestions on the merge issue? I thought data.table roll nearest would do the trick but haven’t had much luck.

**Comment by juthzi** [comment]
Thank you @tjmahr for your feedback. I looked into vroom, but did not find out, how it could help me with this specific problem. 

Thank you @jgeller112 for trying the merge with data.table roll nearest. In theory it looks like a promising function for this issue, but if you did not have any luck, I will not go down the same road. Do you know any other functions like it that work in a similar fashion?

**Comment by juthzi** [comment]
For anyone interested, this is my solution:
I now downsampled the data by first taking out all message-info-rows and in the resulting dataframe only keeping alive each fourth row that contains gaze data. Then, I concatenated back together the message- and non-message-rows and sorted them back into their original order.
Then, I run an if-loop which is far from elegant but does the job. It iterates over the rows and for each row with a message it throws away the next-nearest row or the other one, if the next-nearest is occupied by a message-row.
